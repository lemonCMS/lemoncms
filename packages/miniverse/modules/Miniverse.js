import {
  concat,
  ReplaySubject,
  zip,
} from 'rxjs';
import { ApiService } from './ApiService';
import { EventService } from './EventService';

class Miniverse {
  /**
   *
   * @type {{}}
   */
  services = {};

  /**
   *
   * @type concat
   * */
  concat = concat;

  /**
   *
   * @type {{}}
   */
  helpers = {};

  /**
   *
   * @type {ApiService}
   */
  apiService = null;

  /**
   *
   * @type {EventService}
   */
  eventService = null;

  /**
   *
   * @param services
   * @param apiService
   */
  constructor(services, apiService = null) {

    if (apiService === null) {
      this.apiService = new ApiService()
    } else {
      this.apiService = apiService;
    }

    this.eventService = new EventService;

    this.services = this.objectMap(services, Service => {
      const service = new Service(this, this.apiService, this.eventService);
      if (typeof service.init === 'function') {
        service.setConfCallback(service.init());
      }
      return service;
    });
  }

  /**
   *
   * @param name
   * @returns {boolean}
   */
  isService(name) {
    return typeof this.services[name] !== 'undefined';
  }

  /**
   *
   * @returns {*}
   */
  eject = () => {
    const subjects = {};
    Object.keys(this.services).forEach(service => {
      subjects[service] = this.services[service].eject();
    });
    return this._waitAllDone(subjects);
  };

  /**
   *
   * @returns {*}
   */
  insert = (data) => {
    const subjects = {};
    Object.keys(data).forEach(service => {
      if (this.isService(service)) {
        subjects[service] = this.services[service].insert(data[service]);
      }
    });
    return this._waitAllDone(subjects);
  };

  /**
   *
   * @param name
   * @returns {boolean|*}
   */
  getService = name => {
    if (typeof this.services[name] === 'undefined') {
      throw Error(`Service ${name} is not registered in services`);
    }
    return this.services[name];
  };

  /**
   *
   * @returns {EventService}
   */
  getEventService() {
    return this.eventService;
  }

  /**
   *
   * @param helpers
   */
  setHelpers = helpers => {
    const { client, ...rest } = helpers;
    this.client = client;
    this.helpers = rest;
  };

  /**
   *
   * @param name
   * @returns {boolean|*}
   */
  getHelper = name => {
    if (typeof this.helpers[name] === 'undefined') {
      console.error(`Helper ${name} is not registered in helpers`);
      return false;
    }
    return this.helpers[name];
  };

  /**
   *
   * @param object
   * @param mapFn
   * @returns {{}}
   */
  objectMap = (object, mapFn) => {
    return Object.keys(object).reduce((result, key) => {
      result[key] = mapFn(object[key]);
      return result;
    }, {});
  };

  _waitAllDone = subjects => {
    const subject = new ReplaySubject();
    const toExport = Object.keys(subjects);
    let nextAfter = toExport.length;
    const preparedExport = {};

    if (nextAfter === 0) {
      subject.next({});
      subject.complete();
    }

    toExport.forEach(key => {
      subjects[key].subscribe(data => {
        preparedExport[key] = data;
        nextAfter -= 1;
        if (nextAfter <= 0) {
          subject.next(preparedExport);
          subject.complete();
        }
      });
    });
    return subject;
  };

  zip(...rest) {
    const toZip = rest.map(apiCall => {
      if (typeof apiCall === 'object' && typeof apiCall.subject === 'function') {
        return apiCall.subject();
      }
      return apiCall;
    });
    return zip(...toZip);
  }
}

export default Miniverse;
