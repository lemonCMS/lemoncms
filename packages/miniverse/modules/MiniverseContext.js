import { createContext } from 'react';

const MiniverseContext = createContext({});

























export default MiniverseContext;


