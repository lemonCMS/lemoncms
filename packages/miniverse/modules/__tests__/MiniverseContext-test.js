import React from "react";
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Decorated from './components/DecoratedComponent';
import PassedThroughComponent from './__mock__/PassedThroughComponent';
import MiniverseContext from '../MiniverseContext';
import Miniverse from '../Miniverse';
import Github from './__mock__/Github-mock';
import mockTinyWarning from 'tiny-warning';
import { ajax } from 'rxjs/ajax';
import {
  of,
  throwError
} from 'rxjs';
import { tap } from 'rxjs/operators';

jest.mock("rxjs/ajax");

jest.mock("tiny-warning");
let sP = null;
beforeEach(() => {
  sP = new Miniverse({ Github });
  jest.resetAllMocks();
});

const Mock = () => (
  <MiniverseContext.Provider value={{ miniverse: sP }}>
    <Decorated />
  </MiniverseContext.Provider>);

const PassedThroughMock = () => (
  <MiniverseContext.Provider value={{ miniverse: sP }}>
    <PassedThroughComponent />
  </MiniverseContext.Provider>
);

describe('Miniverse', () => {

  test("Should just render child component", (done) => {
    const tree = renderer
      .create(<PassedThroughMock />);
    process.nextTick(() => {
      const output = tree.toJSON();
      expect(output).toMatchSnapshot();
      done();
    });
  });

  test("Should just log error", (done) => {
    ajax.mockImplementationOnce(() =>
      throwError('error')
    );


    const tree = renderer
      .create(<Mock />);
    process.nextTick(() => {
      sP.getService('Github').getUsers();
      expect(ajax).toHaveBeenCalledTimes(1);
      process.nextTick(() => {
        const output = tree.toJSON();
        expect(output).toMatchSnapshot();
        expect(mockTinyWarning).toBeCalled();
        done();
      });
    });
  });

  test("SJust some stuff to get to 100% store.js", (done) => {
    expect(sP.getService('Github').getSubscription('mock1', 'mock2')).toBeNull();
    sP.getService('Github').importSubject('xxxx');
    sP.getService('Github').importSubject('xxxx');
    done();
  });

  test("Should fetch, watch and render Github users", (done) => {
    ajax.mockImplementationOnce(() =>
      of({
        response: { data: { github: true } }
      })
    );

    sP.getService('Github').getUsers().subscribe(() => {
    });

    const component = mount(<Mock />);

    expect(ajax).toHaveBeenCalledTimes(1);

    process.nextTick(() => {
      const subjects = sP.getService('Github').getLocations();
      expect(Object.keys(sP.getService('Github').getLocations())).toHaveLength(1);
      Object.keys(subjects).forEach(resource => {
        expect(Object.keys(subjects[resource])).toHaveLength(2)
      });


      const wrapped = component.find('TestComponent');
      // console.log(component.debug());
      expect(wrapped).toBeDefined();
      expect(wrapped.props().github.users.data.github).toBeTruthy();
      // expect(component.find('span').text()).toEqual("1");
      process.nextTick(() => {
        component.unmount();
        process.nextTick(() => {
          expect(Object.keys(subjects)).toHaveLength(1);
          Object.keys(subjects).forEach(resource => {
            expect(Object.keys(subjects[resource])).toHaveLength(1)
          });
          done();
        });
      });
    });
  });

  test("Should fetch, watch and render and update Github users", (done) => {
    ajax
      .mockImplementationOnce(() =>
        of({ response: { data: { github: true } } })
      )
      .mockImplementationOnce(() =>
        of({ response: { data: { github: false } } })
      )
    ;

    console.log = jest.fn();
    const sub$ = sP.getService('Github').getUsers().pipe(tap(console.log)).subscribe();

    const component = mount(<Mock />);
    expect(ajax).toHaveBeenCalledTimes(1);

    process.nextTick(() => {
      expect(console.log).toHaveBeenCalled();
      const subjects = sP.getService('Github').getLocations();
      expect(Object.keys(sP.getService('Github').getLocations())).toHaveLength(1);
      Object.keys(subjects).forEach(resource => {
        expect(Object.keys(subjects[resource])).toHaveLength(2)
      });

      let wrapped = null;
      wrapped = component.find('TestComponent');
      // console.log(component.debug());
      expect(wrapped).toBeDefined();
      expect(wrapped.props().github.users.data.github).toBeTruthy();

      sP.getService('Github').getUsers();
      expect(ajax).toHaveBeenCalledTimes(2);

      process.nextTick(() => {
        component.update();
        wrapped = component.find('TestComponent');
        expect(wrapped.props().github.users.data.github).toEqual(false);

        sP.getService('Github').resetUsers();
        expect(ajax).toHaveBeenCalledTimes(2);
        process.nextTick(() => {
          component.update();
          wrapped = component.find('TestComponent');
          expect(wrapped.props().github.users).toEqual(null);
          sub$.unsubscribe();
          done();
        });
      });
    });
  });

  test("Should fetch from cache boolean", () => {
    ajax
      .mockImplementationOnce(() =>
        of({ response: { data: { github: true } } })
      );

    sP.getService('Github').getUsersCacheBoolean();
    sP.getService('Github').getUsersCacheBoolean();
    expect(ajax).toHaveBeenCalledTimes(1);

  });

  test("Should fetch from cache object", () => {
    ajax
      .mockImplementation(() =>
        of({ response: { data: { page: 1 } } })
      )
      .mockImplementation(() =>
        of({ response: { data: { page: 2 } } })
      );

    sP.getService('Github').getUsersCacheObject({ page: 1 });
    sP.getService('Github').getUsersCacheObject({ page: 1 });
    sP.getService('Github').getUsersCacheObject({ page: 2 });
    expect(ajax).toHaveBeenCalledTimes(2);

  });
});
