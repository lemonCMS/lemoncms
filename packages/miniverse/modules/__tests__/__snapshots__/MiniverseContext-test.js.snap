// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Miniverse Should fetch, watch and render Github users 1`] = `
<div>
  Rerenderd: 
  1
   
  {"users":{"data":{"github":true}}}
</div>
`;

exports[`Miniverse Should just log error 1`] = `
<div>
  Rendered: 
  <span>
    2
  </span>
   
  {"users":"error"}
</div>
`;

exports[`Miniverse Should just render child component 1`] = `
<div>
  Rerenderd: 
  1
</div>
`;
